middle_end/flambda2/types/meet_and_join.ml

```diff
@@ -159,439 +438,451 @@ and meet0 env (t1 : TG.t) (t2 : TG.t) : TG.t * TEE.t =
+          | Bottom -> Bottom
+          | Unknown ->
+            (* vlaviron: If the meet of the kinds is Unknown, then both inputs
+               had Unknown kinds. I don't see how we could end up with an array
+               type where the contents are known but we don't know the kind, but
+               in that case we wouldn't be able to call meet because the two
+               sides may have different kinds. So we'll just return the first
+               input, which is guaranteed to be a correct approximation of the
+               meet. *)
+            Ok (Left_input, env)
+          | Ok _ ->
+            map_result
+              ~f:(fun fields : TG.array_contents -> Immutable { fields })
+              (meet_array_of_types env fields1 fields2
+                 ~length:(Array.length fields1))))
     ~contents_is_bottom:(fun (array_contents : TG.array_contents) ->
       match array_contents with
       | Mutable -> false
-      | Immutable { fields } -> List.exists TG.is_obviously_bottom fields)
+      | Immutable { fields } -> Array.exists TG.is_obviously_bottom fields)
     env array_contents1 array_contents2
 
 and meet_variant env ~(blocks1 : TG.Row_like_for_blocks.t Or_unknown.t)
     ~(imms1 : TG.t Or_unknown.t)
     ~(blocks2 : TG.Row_like_for_blocks.t Or_unknown.t)
     ~(imms2 : TG.t Or_unknown.t) :
-    (TG.Row_like_for_blocks.t Or_unknown.t * TG.t Or_unknown.t * TEE.t)
-    Or_bottom.t =
-  let blocks =
-    meet_unknown meet_row_like_for_blocks
-      ~contents_is_bottom:TG.Row_like_for_blocks.is_bottom env blocks1 blocks2
-  in
-  let blocks : _ Or_bottom.t =
-    match blocks with
-    | Bottom | Ok (Or_unknown.Unknown, _) -> blocks
-    | Ok (Or_unknown.Known blocks', _) ->
-      if TG.Row_like_for_blocks.is_bottom blocks' then Bottom else blocks
-  in
-  let imms =
-    meet_unknown meet ~contents_is_bottom:TG.is_obviously_bottom env imms1 imms2
+    (TG.Row_like_for_blocks.t Or_unknown.t * TG.t Or_unknown.t) meet_result =
+  let blocks_meet =
+    { is_bottom =
+        (fun blocks ->
+          match (blocks : _ Or_unknown.t) with
+          | Unknown -> false
+          | Known blocks -> TG.Row_like_for_blocks.is_bottom blocks);
+      mk_bottom = (fun () -> Or_unknown.Known TG.Row_like_for_blocks.bottom);
+      meet =
+        meet_unknown meet_row_like_for_blocks
+          ~contents_is_bottom:TG.Row_like_for_blocks.is_bottom
+    }
   in
-  let imms : _ Or_bottom.t =
-    match imms with
-    | Bottom | Ok (Or_unknown.Unknown, _) -> imms
-    | Ok (Or_unknown.Known imms', _) ->
-      if TG.is_obviously_bottom imms' then Bottom else imms
+  let immediates_meet =
+    { is_bottom =
+        (fun immediates ->
+          match (immediates : _ Or_unknown.t) with
+          | Unknown -> false
+          | Known imms -> TG.is_obviously_bottom imms);
+      mk_bottom = (fun () -> Or_unknown.Known TG.bottom_naked_immediate);
+      meet = meet_unknown meet ~contents_is_bottom:TG.is_obviously_bottom
+    }
```

Comment:
This is going to allocate quite a lot: the record and all the closures.
The functions from the `let rec` are coerced and this currently prevents lifting. @lukemaurer Is there any plan to lift that restriction ?

Reply:
Otherwise we might want to do the lifting by hand. That would produce a `let rec` with both values and functions

Reply:
Should be fixed by inlining `pairwise_dusjunction_meet`